Fixing Building the Documentation Guide (#3757)
owncloud/docs: docs/build-the-docs.md

To [install yarn](https://yarnpkg.com/lang/en/docs/install) following the installation instructions for your operating system.

## Install Build Dependencies

Before you can build the documentation, you must install Antora's dependencies.

> To install them, you just need to run `yarn install`, from the command line in the root directory of the `docs` directory.

Reviewer: ```suggestion
To install them, you just need to run `yarn install` on the command line at the top level of the `docs` directory.
```

This will install all the dependencies specified in `package.json`, which is located in the root directory of the `docs` directory.

It is recommended that you regularly run `yarn install` as from time to time packages are bumped to newer versions.

To generate the documentation in PDF format locally, you need to have `asciidoctor-pdf`. To install `asciidoctor-pdf`, please refer to the [official installation instructions](https://asciidoctor.org/docs/asciidoctor-pdf/#getting-started) or by typing:

```
sudo gem install asciidoctor-pdf
```

You can check the location and version of `asciidoctor-pdf` by invoking following commands:

```
whereis asciidoctor-pdf
asciidoctor-pdf: /usr/local/bin/asciidoctor-pdf

asciidoctor-pdf --version
Asciidoctor PDF 1.6.0 using Asciidoctor 2.0.12 [https://asciidoctor.org]
Runtime Environment (ruby 2.5.1p57 (2018-03-29 revision 63029) [x86_64-linux-gnu]) (lc:UTF-8 fs:UTF-8 in:UTF-8 ex:UTF-8)
```

With the dependencies installed, you are now ready to build (generate) the ownCloud documentation.

## Prepare Your Browser

> It is often helpful to see changes made how they will render quickly. Therefore you can install a plugin for your browser to render .adoc files. You may use the `Asciidoctor.js Live Preview` or any other that is available for your browser - just search and install a suitable one. Post installing, check that accessing local files in the plugin settings is allowed.

Reviewer: ```suggestion
It is very helpful to see how changes to a section will render. Therefore you can install a plugin for your browser to render .adoc files. You may use the `Asciidoctor.js Live Preview` or any other that is available for your browser - just search and install a suitable one. Post installing, check that accessing local files in the plugin settings is allowed.
```

Note, that rendering in the browser will not reslove global variables declared in eg `site.yml` or references to other .adoc files properly. The result shown in the browser may therefore look slightly different to a version that is built via ` yarn antora-local`, but is a good start to catch first typos.  

## Prepared Yarn Commands

> To see all prepared yarn commands run following command:

Reviewer: ```suggestion
To see all, prepared yarn commands run the following command:
```

Using Yarn, as in the example below, is the easiest way to build the documentation. This project has a predefined target (`antora`) which calls Antora, supplying all of the required options to build the docs, to build the documentation on any branch of the [ownCloud documentation repository](https://github.com/owncloud/docs).

> Note that the build process is essential as it must run error free for a valid documentation. If you push an erroring change, CI will complain in the Pull request and disallow any merging. For a quick view on the changes made and without having a full build, you can open the changed file in the browser and view it with the installed pluging which helps finding typos and/or rendering issues quickly. 

Reviewer: ```suggestion
Note that the build process is essential as it must run error free for a valid documentation. If you push a change with errors, the CI will complain in the Pull request and disallow any merging. For a quick view on the changes made and without having a full build, you can open the changed file in the browser and view it with the installed plug-in which helps finding typos and/or rendering issues quickly. 
```

Status: Downloaded newer image for owncloudci/nodejs:11
```

> ## Viewing The HTML Documentation

Reviewer: ```suggestion
## Viewing the HTML Documentation
```

This starts a simple webserver, using the `public` directory, (re)generated by `antora`, as the document root, listening on `http://localhost:8080`. Open the URL in your browser of choice and you'll see two links, as below.

**NOTE:** You may get a different screen telling in the main middle frame "Page not found". In this case, you have to select the manual to view. You find the selectable list on the left side. For owncloud Server, select _master_ as you usually start working there.

![Viewing the locally generated Antora documentation](./images/viewing-the-locally-generated-antora-documentation.png)

> If you're happy with your changes, as with any other change, create a set of meaningful commits and push them to the remote repository. If you're _not_ satisfied with the changes, however, continue to make further updates and start the build process again. Your changes will be reflected in the local version of the site that Serve is rendering.

Reviewer: ```suggestion
If you're happy with your changes, create a set of meaningful commits and push them to the remote repository. If you're _not_ satisfied, continue to work on the file(s) and start the build process again. Your changes will be reflected in the local version of the site that Serve is rendering.
```

We hope that you can see that contributing to the documentation using Antora is a pretty straight-forward process, and not _that_ demanding.

## Generating PDF Documentation

> Run the command below in the root directory of the repository, to generate PDF versions of the _administration_, _developer_ and _user_ manuals.

Reviewer: ```suggestion
Run the command below in the top-level directory of the repository to generate PDF versions of the _administration_, _developer_ and _user_ manuals.
```

- That an error was found
- The file it was found in
- The line of that file where it is located

## TIPS

### Additional Command Line Parameters

You can add additional parameters to the current defined ones. For example, defining the default URL
or additional global attributes. Just add them after the `yarn antora` command. 

### Overwrite the Default URL
If you want to serve your changes locally, you have to overwrite the default URL, which points to https://doc.owncloud.com. You can append a custom URL to the command like this:

```
yarn antora --url http://localhost:8080
 or use
yarn antora-local
```

> Overwriting the default URL to local is especially helpful, if you further want to check for broken links.

Reviewer: ```suggestion
Overwriting the default URL to local is especially helpful if you also want to check for broken links.
```

> ### Attribute Error Searching and Fixing

Reviewer: ```suggestion
### Searching and Fixing Attribute Errors
```

It is very beneficial to use command line attributes when searching and fixing attribute errors. This can be
necessary when you get warnings like: `WARNING: skipping reference to missing attribute: <attribute-name>`

> - First, you may want to search if the attribute-name is used as an attribute all. Run in docs root\

Reviewer: ```suggestion
- First, you may want to check if the attribute name is used as an attribute at all. Run at the top level of the docs repo:
```

`grep -rn --exclude-dir={public,.git,node_modules} \{attribute-name`\
If found, check if the attribute definition is made or passed or needs exclusion. 

> - Second, if no result is found, it may be the case that the erroring attribute is not in the master

Reviewer: ```suggestion
- If no result is found, it may be the case that the error-causing attribute is not in the master
```

branch but in another one. This can be identified by adding a custom attribute to the yarn antora command like:\
`--attribute the-erroring-attribute=HUGO` where HUGO can be anything that is not used and easy to grep.

> - Finally, run in `public` directory: `grep -rn HUGO`. You will see exactly in which branch and file the issue occurs.

Reviewer: ```suggestion
- Finally, run in the `public` directory: `grep -rn HUGO`. You will see exactly in which branch and file the issue occurs.
```
